# Exercise 0


Welcome to the Data Science Exercise!
Since this is the first week, and the first exercise group takes place immediately after the first lecture, we will go over basic `pandas` in this first Exercise.

If there is something from the first lecture that you want to hear more about, please tell us! We will prioritize these topics and then move on to this `pandas` notebook. Of course you can still ask questions about the lecture while we are doing the `pandas` notebook!

In [1]:
!pip install -r requirements.txt

ERROR: Ignored the following versions that require a different python version: 2.1.0 Requires-Python >=3.10; 2.1.0rc1 Requires-Python >=3.10; 2.1.1 Requires-Python >=3.10
ERROR: Could not find a version that satisfies the requirement numpy==2.1.1 (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5, 1.20.0, 1.20.1, 1.20.2, 1.20.3, 1.21.0, 1.21.1, 1.21.2, 1.21.3, 1.21.4, 1.21.5, 1.21.6, 1.22.0, 1.22.1, 1.22.2, 1.22.3, 1.22.4, 1.23.0, 1.23.1, 1.23.2, 1.23.3, 1.23.4, 1.23.5, 1.24.0, 1.24.1, 1.24.2, 1.24.3, 1.24

# The Series Data Structure

The `Series` is one of the core data structures in `pandas`. You think of it a cross between a list and a dictionary. The items are all stored in an order and there's labels with which you can retrieve them.

In [5]:
import pandas as pd

print(pd.__version__)

2.2.3


Jupyter Notebook provides an extended help and show the documentation, using the question mark operator `?`

**Note**: Depending on your IDE this may look different or appear in a different place. You can of course always use the official documentation as well. Especially for `pandas` it's very good and comprehensive.

In [6]:
# pd.Series?

In [7]:
animals = ['Tiger', 'Bear', 'Moose']
animal_series = pd.Series(animals)
animal_series

0    Tiger
1     Bear
2    Moose
dtype: object

In [8]:
numbers = [1, 2, 3]
pd.Series(numbers)

0    1
1    2
2    3
dtype: int64

`pandas` stores series values in a typed array using the `NumPy` library. This offers significant speed-up when processing data versus traditional Python lists.

There's some other typing details that exist for performance that are important to know. The most important is how `NumPy` and thus `pandas` handle missing data. In Python, we have the `None` type to indicate a lack of data. But what do we do if we want to have a typed list like we do in the series object?

In [7]:
animals = ['Tiger', 'Bear', None]
pd.Series(animals)

0    Tiger
1     Bear
2     None
dtype: object

 If we create a list of numbers, integers or floats, and put in the `None` type, `pandas` automatically converts this to a special floating point value designated as `NaN`, which stands for 'not a number'.

In [9]:
numbers = [1, 2, None]
pd.Series(numbers)

0    1.0
1    2.0
2    NaN
dtype: float64

`NaN` is not `None` and when we try the equality test, it evaluates to `False`.

In [10]:
import numpy as np

# noinspection PyComparisonWithNone
np.nan == None

False

It turns out that you also can't do an equality test of `NaN` to itself. When you do, the answer is always `False`.

Why is explained [here](https://stackoverflow.com/questions/1565164/what-is-the-rationale-for-all-comparisons-returning-false-for-ieee754-nan-values/1573715).

In [11]:
np.nan == np.nan

False

 You need to use special functions to test for the presence of not a number, such as the Numpy library `isnan()`.

In [12]:
np.isnan(np.nan)

np.True_

A series can be created from dictionary data.

In [14]:
sports = {
    'Football': 'Germany',
    'Golf': 'Scotland',
    'Sumo': 'Japan',
    'Taekwondo': 'South Korea'
}
sports["Golf"]

'Scotland'

In [15]:
sports = {
    'Football': 'Germany',
    'Golf': 'Scotland',
    'Sumo': 'Japan',
    'Taekwondo': 'South Korea'
}
s = pd.Series(sports)
s

Football         Germany
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

In [16]:
s.index

Index(['Football', 'Golf', 'Sumo', 'Taekwondo'], dtype='object')

You could also separate your index creation from the data by passing in the index as a list explicitly to the series.

In [17]:
s = pd.Series(['Tiger', 'Bear', 'Moose'], index=['India', 'America', 'Canada'])
s

India      Tiger
America     Bear
Canada     Moose
dtype: object

In the following example, we pass in a dictionary of four items but only two are preserved in the `Series` object because of the `index` argument. We see that `"Hockey"` is included in the `index` argument, but because it does not exist in the dictionary, it has no value associated with it.

In [18]:
sports = {
    'Football': 'Germany',
    'Golf': 'Scotland',
    'Sumo': 'Japan',
    'Taekwondo': 'South Korea'
}
s = pd.Series(sports, index=['Golf', 'Sumo', 'Hockey'])
s

Golf      Scotland
Sumo         Japan
Hockey         NaN
dtype: object

# Querying a Series

A `pandas` `Series` can be queried, either by the index position or the index label. As we saw, if you don't give an index to the series, the position and the label are effectively the same values. To query by numeric location, starting at zero, use the `iloc` attribute. To query by the index label, you can use the `loc` attribute.

In [19]:
import pandas as pd

In [20]:
sports = {
    'Football': 'Germany',
    'Golf': 'Scotland',
    'Sumo': 'Japan',
    'Taekwondo': 'South Korea'
}
s = pd.Series(sports)
s

Football         Germany
Golf            Scotland
Sumo               Japan
Taekwondo    South Korea
dtype: object

Access a row by integer index

In [21]:
s.iloc[3]

'South Korea'

Access a row by label

In [22]:
s.loc['Golf']

'Scotland'

Keep in mind that `iloc` and `loc` are not methods, they are attributes. So you don't use parentheses to query them, but square brackets instead, which we'll call the indexing operator.

If you pass in an integer parameter, the operator will behave as if you want it to query via the `iloc` attribute.

If your `pandas` version is current enough, you should see a warning:

`/tmp/ipykernel_144448/3028526611.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use ser.iloc[pos]`

This means we can currently do it like this, but in a future `pandas` version the ability to do so will be removed and we should use `.iloc` instead.

In [23]:
s[3]

/var/folders/9j/1050b49n4kj_615mds_fqt2c0000gn/T/ipykernel_97578/3028526611.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  s[3]


'South Korea'

If you pass in an object, it will query as if you wanted to use the label based `loc` attribute.

In [24]:
s['Golf']

'Scotland'

Here's an example using some new sports data, where countries are indexed by integer.

In [4]:
sports = {
    99: 'Germany',
    100: 'Scotland',
    101: 'Japan',
    102: 'South Korea'
}
s = pd.Series(sports)

If we try and call `s[0]`, we get a `KeyError`.

In [26]:
# s[0]  # we get a key error, because there's no item in the sports list with an index of zero.

In [27]:
s.iloc[0]  # Instead we have to call iloc explicitly if we want the first item.

'Germany'

The first entry has the index `99`. This may look confusing - we only have four elements, so how can the element have such a high index?

That's because we have specified the index when we created the `Series` object: the keys in our `sports` dictionary! It works essentially the same way here as it does in a dictionary: we associate the index `99` with the value `'Germany'`.

In [28]:
s[99]

'Germany'

Let's talk about working with the data.

A data Series of floating point values.

In [29]:
s = pd.Series([100.00, 120.00, 101.00, 3.00])
s

0    100.0
1    120.0
2    101.0
3      3.0
dtype: float64

An iteration over all the items in the series and adds them together to get a total.

In [30]:
total = 0
for item in s:
    total += item
print(total)

324.0


This works, but it's slow.

We can do it much faster by using `Numpy`'s `sum` method:

In [31]:
import numpy as np

total = np.sum(s)
print(total)

324.0


This `Series` is a bit small. Let's try it with a bigger one. First we create a new `Series` object with a lot of (random) numbers:

In [32]:
s = pd.Series(np.random.randint(0, 1000, 10000))  # low=0, high=1000, size=10000
s.head()

0    539
1    470
2    228
3    253
4    356
dtype: int64

In [33]:
len(s)

10000

The function we're going to use is called `timeit`; this function will run our code a few times to determine, on average, how long it takes.

In [34]:
%%timeit -n 100
summary = 0
for item in s:
    summary += item

645 µs ± 24 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Now let's try again with the `sum` function from `NumPy`:

In [35]:
%%timeit -n 100
summary = np.sum(s)

15.7 µs ± 1.55 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


This is much faster!

A related feature in Pandas and NumPy is called **broadcasting**. With broadcasting, you can apply an operation to every value in the series, changing the series.

In [36]:
s.head()

0    539
1    470
2    228
3    253
4    356
dtype: int64

In [37]:
s += 2  # adds two to each item in s using broadcasting

In [38]:
s.head()

0    541
1    472
2    230
3    255
4    358
dtype: int64

In [39]:
for label, value in s.items():
    s.at[label] = value + 2
s.head()

0    543
1    474
2    232
3    257
4    360
dtype: int64

Let's try and time the two approaches for increasing the value

In [40]:
%%timeit -n 10
s = pd.Series(np.random.randint(0, 1000, 10000))
for label, value in s.items():
    s.loc[label] = value + 2

147 ms ± 2.4 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [ ]:
%%timeit -n 10
s = pd.Series(np.random.randint(0, 1000, 10000))
s += 2

The approach using broadcasting is significantly faster, shorter and more readable.

One last note on using the indexing operators to access series data. The `.loc` attribute lets you not only modify data in place, but also add new data as well.

Pandas will automatically change the underlying NumPy types as appropriate.

In [ ]:
s = pd.Series([1, 2, 3])

Look at the `dtype`: It says `int64`.

In [ ]:
s

Now check the `dtype` after we add a string to the `Series`:

In [ ]:
s.loc['Animal'] = 'Dog'
s

It has changed to `object`.

Revisiting the issue of countries and their national sports, it turns out that many countries seem to like this game cricket. We go back to our original series on sports. It's possible to create a new `Series` object with multiple entries for cricket and merge it with the original by using the `pandas` `concat` function.

In [ ]:
original_sports = pd.Series({
    'Football': 'Germany',
    'Golf': 'Scotland',
    'Sumo': 'Japan',
    'Taekwondo': 'South Korea'
})

cricket_loving_countries = pd.Series([
    'Australia',
    'Barbados',
    'Pakistan',
    'England'
], index=[
    'Cricket',
    'Cricket',
    'Cricket',
    'Cricket'
])

all_countries = pd.concat([original_sports, cricket_loving_countries])

In [ ]:
all_countries

Finally, we see that when we query the appended series for those who have cricket as their national sport, we don't get a single value, but another `Series`.

In [ ]:
all_countries.loc['Cricket']

# The `DataFrame` Data Structure

The `DataFrame` data structure is the heart of the `pandas` library. It's a primary object that you'll be working with in data analysis and cleaning tasks.

The `DataFrame` is conceptually a two-dimensional `Series` object, where there's an index and multiple columns of content, with each column having a label. In fact, the distinction between a column and a row is really only a conceptual distinction. And you can think of the `DataFrame` itself as simply a two-axes labeled array.

## `DataFrame` creation
You can create a `DataFrame` in many different ways, some of which you might expect. For instance, you can use multiple `Series`, where each `Series` represents a row of data. Or you could use multiple dictionaries, where each dictionary represents a row of data.

We are going to create three purchase order records as `Series` objects for a sort of fictional store.

In [1]:
import pandas as pd

purchase_1 = pd.Series({
    'Name': 'Matthias',
    'Item Purchased': 'Dog Food',
    'Cost': 22.50
})
purchase_2 = pd.Series({
    'Name': 'Thomas',
    'Item Purchased': 'Kitty Litter',
    'Cost': 2.50
})
purchase_3 = pd.Series({
    'Name': 'Christina',
    'Item Purchased': 'Bird Seed',
    'Cost': 5.00
})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

In this example, we wanted to select data associated with Store 2, so we query the `loc` attribute with one parameter.

## Data Access
Similar to how we did it when working with a `Series`, we can extract data using the `iloc` and `loc` attributes when working with a `DataFrame`.



In [2]:
df.iloc[0]

Name              Matthias
Item Purchased    Dog Food
Cost                  22.5
Name: Store 1, dtype: object

Because the `DataFrame` is two-dimensional, passing a single value to the `loc` indexing operator will return a series if there's only one row to return.

In the following example, we want to select data associated with Store 2, so we query the `loc` attribute with one parameter.

In [3]:
df.loc['Store 2']

Name              Christina
Item Purchased    Bird Seed
Cost                    5.0
Name: Store 2, dtype: object

 We can check the data type of the return using the Python `type` function.

In [4]:
type(df.loc['Store 2'])

pandas.core.series.Series

It's important to remember that the indices and column names along either axes, horizontal or vertical, could be non-unique.

For instance, in this example, we see two purchase records for Store 1 as different rows.

In [5]:
df.loc['Store 1']

Name Item Purchased  Cost
Store 1  Matthias       Dog Food  22.5
Store 1    Thomas   Kitty Litter   2.5

We see that Matthias and Thomas both shop at the same pets supply store.

In this case, the return value is a `DataFrame`:

In [6]:
type(df.loc['Store 1'])

pandas.core.frame.DataFrame

One of the powers of the `pandas` `DataFrame` is that you can quickly select data based on multiple axes.

For instance, if you wanted to just list the costs for Store 1, you would supply two parameters to `.loc`, one being the row **index** and the other being the **column** name.

In [7]:
df.loc['Store 1', 'Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

## Column Selection

What if we just wanted to do column selection and just get a list of all the costs? Well, there's a couple of options.

### 1. Transpose `DataFrame`
You can get a transpose of the `DataFrame`, using the capital T attribute, which swaps all the columns and rows.

In [8]:
df.T

Store 1       Store 1    Store 2
Name            Matthias        Thomas  Christina
Item Purchased  Dog Food  Kitty Litter  Bird Seed
Cost                22.5           2.5        5.0

And we can then use `.loc`:

In [9]:
df.T.loc['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: object

This works, but it's kind of ugly.

### 2. Indexing Operator

In a `pandas` `DataFrame`, columns always have a name;  so this selection is always label based.

(For those familiar with relational databases, this operator is analogous to column projection.)

In [10]:
df['Cost']

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

### 3. Operation Chaining
Finally, since the result of using the indexing operators is a `DataFrame` or `Series`, you can chain operations together.

In [11]:
# loc operation then indexing operation
s1 = df.loc['Store 1']
s1['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

In [12]:
# chaining of loc and indexing
df.loc['Store 1']['Cost']

Store 1    22.5
Store 1     2.5
Name: Cost, dtype: float64

But chaining can come with some costs and is best avoided if you can use another approach. In particular, chaining tends to cause `pandas` to return a copy of the `DataFrame` instead of a view on the `DataFrame`. For selecting a data, this is not a big deal, though it might be slower than necessary. If you are changing data though, this is an important distinction and can be a source of error.

### 4. Full slice with `.loc`

As we saw, .loc does row selection, and it can take two parameters, the row index and the list of column names. italso supports slicing. If we wanted to select all rows, we can use a column to indicate a full slice from beginning to end. And then add the column name as the second parameter as a string. In fact, if we wanted to include multiply columns, we could do so in a list. And Pandas will bring back only the columns we have asked for.

In [13]:
lst = [1, 2, 3]
lst[1:]

[2, 3]

In [14]:
lst = [1, 2, 3, 4, 5]
lst[1:3]

[2, 3]

In [15]:
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

In [16]:
df.loc[:, ['Name', 'Cost']]

Name  Cost
Store 1   Matthias  22.5
Store 1     Thomas   2.5
Store 2  Christina   5.0

## Dropping Data

It's easy to delete data in `Series` and `DataFrame`s, and we can use the drop function to do so. This function takes a single parameter, which is the index or roll label, to drop.

In [17]:
df.drop('Store 1')

Name Item Purchased  Cost
Store 2  Christina      Bird Seed   5.0

The drop function doesn't change the `DataFrame` by default. And instead, returns to you a copy of the `DataFrame` with the given rows removed.

We can see that our original `DataFrame` is still intact.

In [18]:
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

Let's make a copy with the copy method and do a drop on it instead.

In [19]:
copy_df = df.copy()
copy_df = copy_df.drop('Store 1')
copy_df

Name Item Purchased  Cost
Store 2  Christina      Bird Seed   5.0

This is a very typical pattern in `pandas`, where in-place changes to a `DataFrame` are only done if need be, usually on changes involving indices.

In [ ]:
# copy_df.drop?

 Drop has two interesting optional parameters.

- The first is called in place, and if it's set to true, the `DataFrame` will be updated in place, instead of a copy being returned.

- The second parameter is the axes, which should be dropped.

    - By default, this value is `0`, indicating the row axes.

    - But you could change it to `1` if you want to drop a column.


There is a second way to drop a column, however. This is directly through the use of the indexing operator, using the `del` keyword.


In [20]:
del copy_df['Name']
copy_df

Item Purchased  Cost
Store 2      Bird Seed   5.0

This way of dropping data, however, takes immediate effect on the `DataFrame` and does not return a view.

## Adding a new column

Finally, adding a new column to the `DataFrame` is as easy as assigning it to some value.

For instance, if we wanted to add a new location as a column with default value of `None`, we could do so by using the assignment operator after the square brackets.

In [21]:
df['Location'] = None
df

Name Item Purchased  Cost Location
Store 1   Matthias       Dog Food  22.5     None
Store 1     Thomas   Kitty Litter   2.5     None
Store 2  Christina      Bird Seed   5.0     None

This broadcasts the default value to the new column immediately.


**In conclusion**, we presented in this chapter: `DataFrame` creation, data access, column selection, dropping of data and adding a new column.

# `Dataframe` Indexing and Loading

The common work flow is to read your data into a `DataFrame` then reduce this `DataFrame` to the particular columns or rows that you're interested in working with. As you've seen, the `pandas` toolkit tries to give you views on a `DataFrame`. This is much faster than copying data and much more memory efficient too.

But it does mean that if you're manipulating the data you have to be aware that any changes to the `DataFrame` you're working on may have an impact on the original `DataFrame`.

Here's an example using our same purchasing `DataFrame` from earlier. We can create a `Series` based on just the cost category using the square brackets.

## Data Index

In [22]:
import pandas as pd

purchase_1 = pd.Series({
    'Name': 'Matthias',
    'Item Purchased': 'Dog Food',
    'Cost': 22.50
})
purchase_2 = pd.Series({
    'Name': 'Thomas',
    'Item Purchased': 'Kitty Litter',
    'Cost': 2.50
})
purchase_3 = pd.Series({
    'Name': 'Christina',
    'Item Purchased': 'Bird Seed',
    'Cost': 5.00
})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  22.5
Store 1     Thomas   Kitty Litter   2.5
Store 2  Christina      Bird Seed   5.0

Indexing Operator:

In [23]:
costs = df['Cost']
costs

Store 1    22.5
Store 1     2.5
Store 2     5.0
Name: Cost, dtype: float64

Then we can increase the cost in this `Series` using broadcasting.

In [24]:
costs += 2
costs

Store 1    24.5
Store 1     4.5
Store 2     7.0
Name: Cost, dtype: float64

Now if we look at our original `DataFrame`, we see those costs have risen as well. This is an important consideration to watch out for. If you want to explicitly use a copy, then you should consider calling the copy method on the `DataFrame` for it first.

In [25]:
df

Name Item Purchased  Cost
Store 1   Matthias       Dog Food  24.5
Store 1     Thomas   Kitty Litter   4.5
Store 2  Christina      Bird Seed   7.0

## Loading of Data

In this course, we'll be largely using smaller, moderate-sized datasets. As I mentioned, a common workflow is to read the dataset in, usually from some external file. We saw previously how you can do this using Python, and lists, and dictionaries. You can imagine how you might use those dictionaries to create a `pandas` `DataFrame`.

Thankfully, `pandas` has built-in support for delimited files such as CSV files as well as a variety of other data formats including relational databases, Excel, and HTML tables.

We have saved a CSV file called `olympics.csv`, which has data from Wikipedia that contains a summary list of the medal various countries have won at the Olympics.

### Reading a CSV

In [26]:
import pandas as pd

We can read this into a `DataFrame` by calling the `read_csv` function of the module. When we look at the `DataFrame` we see that the first cell has an **NaN** in it since it's an empty value, and the rows have been automatically indexed for us.

In [27]:
df = pd.read_csv('olympics.csv')
df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

It seems pretty clear that the first row of data in the `DataFrame` is what we really want to see as the column names. It also seems like the first column in the data is the country name, which we would like to make an index.


`read_csv` has a number of parameters that we can use to indicate to Pandas how rows and columns should be labeled.

For instance, we can use the `index_col` to indicate which column should be the index, and we can also use the header parameter to indicate which row from the data file should be used as the header.

Let's re-import that data and set the index value to be `0` which is the first column and let set a column headers to be read from the second row of data. We can do this by using the `skiprows` parameters, to tell `pandas` to ignore the first row, which was made up of numeric column names.

In [28]:
df = pd.read_csv('olympics.csv', index_col=0, skiprows=1)
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

Now this data came from the all-time [Olympic games medal table](https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table) on Wikipedia.

If we head to the page we could see that instead of running gold, silver and bronze in the pages, these nice little icons with a one, a two, and a three in them.

## Rename Columns

`pandas` stores a list of all the columns in the `.columns` attribute. We can change the values of the column names by iterating over this list and calling the rename method of the `DataFrame`.

In [29]:
df.columns

Index(['№ Summer', '01 !', '02 !', '03 !', 'Total', '№ Winter', '01 !.1',
       '02 !.1', '03 !.1', 'Total.1', '№ Games', '01 !.2', '02 !.2', '03 !.2',
       'Combined total'],
      dtype='object')

Here we just iterate through all the columns looking to see if they start with a `01`, `02`, `03` or numeric character. If they do, we can call rename and set the column parameters to a dictionary with the keys being the *column* we want to replace and the *value* being the new value we want.

Here we slice some of the old values in two, since we don't want to lose the unique appended values. We'll also set the ever-important `inplace` parameter to `True` so `pandas` knows to update this `DataFrame` directly.

In [30]:
for col in df.columns:
    if col[:2] == '01':
        df.rename(columns={col: 'Gold' + col[4:]}, inplace=True)
    if col[:2] == '02':
        df.rename(columns={col: 'Silver' + col[4:]}, inplace=True)
    if col[:2] == '03':
        df.rename(columns={col: 'Bronze' + col[4:]}, inplace=True)
    if col[:1] == '№':
        df.rename(columns={col: '#' + col[1:]}, inplace=True)

df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

## Conclusion
In this section we have discussed a simple workflow to bring the data to a state where we can query it. For this we discussed the topics of *data Index*, *loading of data* and *renaming of columns*.

# Querying a `DataFrame`

## Boolean masking

Before we talk about how to query `DataFrame`s, we need to talk about Boolean masking. Boolean masking is the heart of fast and efficient querying in `NumPy`.

A Boolean mask is an array which can be of one dimension like a `Series`, or two dimensions like a `DataFrame`, where each of the values in the array are either `True` or `False`. This array is essentially overlaid on top of the data structure that we're querying. And any cell aligned with the `True` value will be admitted into our final result, and any cell aligned with a `False` value will not.

Boolean masks are created by applying operators directly to the `pandas` `Series` or `DataFrame` objects.

For instance, in our Olympics data set, you might be interested in seeing only those countries who have achieved a gold medal at the summer Olympics. To build a Boolean mask for this query, we project the gold column using the indexing operator and apply the **greater than operator** with a comparison value of zero. This is essentially broadcasting a comparison operator, greater than, with the results being returned as a Boolean series.

In [31]:
df['Gold'] > 0

Afghanistan (AFG)                               False
Algeria (ALG)                                    True
Argentina (ARG)                                  True
Armenia (ARM)                                    True
Australasia (ANZ) [ANZ]                          True
                                                ...  
Independent Olympic Participants (IOP) [IOP]    False
Zambia (ZAM) [ZAM]                              False
Zimbabwe (ZIM) [ZIM]                             True
Mixed team (ZZX) [ZZX]                           True
Totals                                           True
Name: Gold, Length: 147, dtype: bool

The resulting `Series` is indexed where the value of each cell is either `True` or `False` depending on whether a country has won at least one gold medal, and the index is the country name.

So this builds us the Boolean mask, which is half the battle. What we want to do next is overlay that mask on the `DataFrame`.

## `where` method

We can do this using the `where` method. The `where` method takes a Boolean mask as a condition, applies it to the `DataFrame` or `Series`, and returns a new `DataFrame` or `Series` of the same shape. Let's apply this Boolean mask to our Olympics data and create a `DataFrame` of only those countries who have won a gold at a summer games.

In [32]:
len(df[df["Gold"] > 0])

100

In [33]:
only_gold = df.where(df['Gold'] > 0)
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)             NaN   NaN     NaN     NaN    NaN       NaN   
Algeria (ALG)                12.0   5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)              23.0  18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                 5.0   1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]       2.0   3.0     4.0     5.0   12.0       0.0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)           NaN       NaN       NaN      NaN      NaN     NaN   
Algeria (ALG)               0.0       0.0       0.0      0.0     15.0     5.0   
Argentina (ARG)             0.0       0.0       0.0      0.0     41.0    18.0   
Armenia (ARM)               0.0       0.0       0.0      0.0     11.0     1.0   
Australasia (ANZ) [ANZ]     0.0       0.0       0.0      0.0      2.0     3.0   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)             NaN       NaN             NaN  
Algeria (ALG)                 2.0       8.0            15.0  
Argentina (ARG)              24.0      28.0            70.0  
Armenia (ARM)                 2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       4.0       5.0            12.0

We see that the resulting `DataFrame` keeps the original indexed values, and only data from countries that met the condition are retained. All the countries which did not meet the condition have `NaN` data instead. This is okay. Most statistical methods built into the `DataFrame` object ignore values of `NaN`.

## `count` method

For instance, if we call `.count()` on the `only_gold` `DataFrame`:

In [34]:
only_gold['Gold'].count()

np.int64(100)

We see that there are 100 countries which have had gold medals awarded at the summer games,

While if we call `.count()` on the original `DataFrame`, we see that there are 147 countries total.

In [35]:
df['Gold'].count()

np.int64(147)

## Drop rows without data

Often we want to drop those rows which have no data. To do this, we can use the `dropna()` method.

You can optionally provide `dropna()` the axes it should be considering. Remember that the `axes` argument is just an indicator for the columns or rows and that the default value for the argument is zero, which means rows.

In [36]:
only_gold = only_gold.dropna()
only_gold.head()

# Summer   Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                  12.0    5.0     2.0     8.0   15.0       3.0   
Argentina (ARG)                23.0   18.0    24.0    28.0   70.0      18.0   
Armenia (ARM)                   5.0    1.0     2.0     9.0   12.0       6.0   
Australasia (ANZ) [ANZ]         2.0    3.0     4.0     5.0   12.0       0.0   
Australia (AUS) [AUS] [Z]      25.0  139.0   152.0   177.0  468.0      18.0   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                 0.0       0.0       0.0      0.0     15.0   
Argentina (ARG)               0.0       0.0       0.0      0.0     41.0   
Armenia (ARM)                 0.0       0.0       0.0      0.0     11.0   
Australasia (ANZ) [ANZ]       0.0       0.0       0.0      0.0      2.0   
Australia (AUS) [AUS] [Z]     5.0       3.0       4.0     12.0     43.0   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                 5.0       2.0       8.0            15.0  
Argentina (ARG)              18.0      24.0      28.0            70.0  
Armenia (ARM)                 1.0       2.0       9.0            12.0  
Australasia (ANZ) [ANZ]       3.0       4.0       5.0            12.0  
Australia (AUS) [AUS] [Z]   144.0     155.0     181.0           480.0

## Indexing Operator with Boolean Mask

We do not actually have to use the where function explicitly. The `pandas` developers allow the indexing operator to take a Boolean mask as a value instead of just a list of column names.

Here's a more concise example of how we could query this `DataFrame`. You'll notice that there are no `NaN`s when you query the `DataFrame` in this manner. `pandas` automatically filters out the rows with `NaN` values.

In [37]:
only_gold = df[df['Gold'] > 0]
only_gold.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Algeria (ALG)                    12     5       2       8     15         3   
Argentina (ARG)                  23    18      24      28     70        18   
Armenia (ARM)                     5     1       2       9     12         6   
Australasia (ANZ) [ANZ]           2     3       4       5     12         0   
Australia (AUS) [AUS] [Z]        25   139     152     177    468        18   

                           Gold.1  Silver.1  Bronze.1  Total.1  # Games  \
Algeria (ALG)                   0         0         0        0       15   
Argentina (ARG)                 0         0         0        0       41   
Armenia (ARM)                   0         0         0        0       11   
Australasia (ANZ) [ANZ]         0         0         0        0        2   
Australia (AUS) [AUS] [Z]       5         3         4       12       43   

                           Gold.2  Silver.2  Bronze.2  Combined total  
Algeria (ALG)                   5         2         8              15  
Argentina (ARG)                18        24        28              70  
Armenia (ARM)                   1         2         9              12  
Australasia (ANZ) [ANZ]         3         4         5              12  
Australia (AUS) [AUS] [Z]     144       155       181             480

The syntax might look a little messy, especially if you're not used to programming languages with overloaded operators, but the result is that you're able to filter and reduce data frames relatively quickly.

## Chain Statements

One more thing to keep in mind if you're not used to Boolean or bit-masking for data reduction. The output of two Boolean masks being compared with logical operators is another Boolean mask. This means that you can chain together a bunch of and/or statements in order to create more complex queries, and the result is a single Boolean mask.

For instance, we could create a mask for all of those countries who have received a gold in the summer Olympics and logically order that with all of those countries who have received a gold in the winter Olympics. If we apply this to the `DataFrame` and use the length function to see how many rows there are, we see that there are 101 countries which have won a gold metal at some time.

In [38]:
len(df[(df['Gold'] > 0) | (df['Gold.1'] > 0)])

101

Another example. Have there been any countries who have only won a gold in the winter Olympics and never in the summer Olympics?

Here's one way to answer that.

In [39]:
df[(df['Gold.1'] > 0) & (df['Gold'] == 0)]

# Summer  Gold  Silver  Bronze  Total  # Winter  Gold.1  \
Liechtenstein (LIE)        16     0       0       0      0        18       2   

                     Silver.1  Bronze.1  Total.1  # Games  Gold.2  Silver.2  \
Liechtenstein (LIE)         2         5        9       34       2         2   

                     Bronze.2  Combined total  
Liechtenstein (LIE)         5               9

Our neighbouring country Liechtenstein.

Extremely important, and often an issue for new users, is to remember that each Boolean mask needs to be encased in parentheses because of the order of operations. This can cause no end of frustration if you're not used to it, so be careful.


# Indexing `DataFrame`s

As we have  seen, both `Series` and `DataFrame`s can have indices applied to them. The index is essentially a row level label, and we know that rows correspond to axis zero. In our Olympics data, we indexed the data frame by the name of the country. Indices can either be inferred, such as when we create a new `Series` without an index, in which case we get numeric values, or they can be set explicitly, like when we use the dictionary object to create the series, or when we loaded data from the CSV file and specified the header.

In [40]:
df.head()

# Summer  Gold  Silver  Bronze  Total  # Winter  \
Afghanistan (AFG)              13     0       0       2      2         0   
Algeria (ALG)                  12     5       2       8     15         3   
Argentina (ARG)                23    18      24      28     70        18   
Armenia (ARM)                   5     1       2       9     12         6   
Australasia (ANZ) [ANZ]         2     3       4       5     12         0   

                         Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
Afghanistan (AFG)             0         0         0        0       13       0   
Algeria (ALG)                 0         0         0        0       15       5   
Argentina (ARG)               0         0         0        0       41      18   
Armenia (ARM)                 0         0         0        0       11       1   
Australasia (ANZ) [ANZ]       0         0         0        0        2       3   

                         Silver.2  Bronze.2  Combined total  
Afghanistan (AFG)               0         2               2  
Algeria (ALG)                   2         8              15  
Argentina (ARG)                24        28              70  
Armenia (ARM)                   2         9              12  
Australasia (ANZ) [ANZ]         4         5              12

## Set an Index

Another option for setting an index is to use the `set_index()` function. This function takes a list of columns and promotes those columns to an index. `set_index()` is a **destructive process**, it doesn't keep the current index. If you want to keep the current index, you need to manually create a new column and copy into it values from the `index` attribute.

Let's go back to our Olympics `DataFrame`. Let's say that we don't want to index the `DataFrame` by countries, but instead want to index by the number of gold medals that were won at summer games. First we need to preserve the country information into a new column. We can do this using the indexing operator or the string that has the column label. Then we can use  `set_index()` to set the index of the column to summer gold medal wins.

In [41]:
df['country'] = df.index
df = df.set_index('Gold')
df.head()

# Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  Bronze.1  \
Gold                                                                          
0           13       0       2      2         0       0         0         0   
5           12       2       8     15         3       0         0         0   
18          23      24      28     70        18       0         0         0   
1            5       2       9     12         6       0         0         0   
3            2       4       5     12         0       0         0         0   

      Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
Gold                                                                 
0           0       13       0         0         2               2   
5           0       15       5         2         8              15   
18          0       41      18        24        28              70   
1           0       11       1         2         9              12   
3           0        2       3         4         5              12   

                      country  
Gold                           
0           Afghanistan (AFG)  
5               Algeria (ALG)  
18            Argentina (ARG)  
1               Armenia (ARM)  
3     Australasia (ANZ) [ANZ]

You'll see that when we create a new index from an existing column it appears that a new first row has been added with empty values. This isn't quite what's happening. And we know this in part because an empty value is actually rendered either as a none or an `NaN` if the data type of the column is numeric. What's actually happened is that the index has a name. Whatever the column name was in the Jupyter notebook has just provided this in the output.

## Reset Index

We can get rid of the index completely by calling the function `reset_index()`. This promotes the index into a column and creates a default numbered index.

In [42]:
df = df.reset_index()
df.head()

Gold  # Summer  Silver  Bronze  Total  # Winter  Gold.1  Silver.1  \
0     0        13       0       2      2         0       0         0   
1     5        12       2       8     15         3       0         0   
2    18        23      24      28     70        18       0         0   
3     1         5       2       9     12         6       0         0   
4     3         2       4       5     12         0       0         0   

   Bronze.1  Total.1  # Games  Gold.2  Silver.2  Bronze.2  Combined total  \
0         0        0       13       0         0         2               2   
1         0        0       15       5         2         8              15   
2         0        0       41      18        24        28              70   
3         0        0       11       1         2         9              12   
4         0        0        2       3         4         5              12   

                   country  
0        Afghanistan (AFG)  
1            Algeria (ALG)  
2          Argentina (ARG)  
3            Armenia (ARM)  
4  Australasia (ANZ) [ANZ]

##  United States Census Data

Let's change data sets and look at some census data for a better example. This data is stored in the file `census.csv` and comes from the *United States Census Bureau*. In particular, this is a breakdown of the population level data at the US county level. It's a great example of how different kinds of data sets might be formatted when you're trying to clean them. For instance, in this data set there are two summarized levels, one that contains summary data for the whole **country**. And one that contains summary data for each **state**, and one that contains summary data for each **county**.

In [43]:
df = pd.read_csv('census.csv')
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
0        4779736            4780127          4785161  ...          0.002295   
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
0         -0.193196          0.381066          0.582002         -0.467369   
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
0     1.030015     0.826644     1.383282     1.724718     0.712594  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  

[5 rows x 100 columns]

### Get Unique Values from a Column

We often want to see a list of all the unique values in a given column. In this `DataFrame`, we see that the possible values for the sum level are using the unique function on the `DataFrame`.

Here we can run unique on the `SUMLEV` of our current `DataFrame` and see that there are only two different values, 40 and 50.

In [44]:
df['SUMLEV'].unique()

array([40, 50])

### Boolean Mask on Value

Let's get rid of all the rows that are summaries at the state level and just keep the county data (i.e. level 50).


In [45]:
df = df[df['SUMLEV'] == 50]  # 50 is the level for the county data
df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   
5          57322              57322            57373  ...          1.807375   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   
5         -1.177622         -1.748766         -2.062535         -1.369970   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  
5     1.859511    -0.848580    -1.402476    -1.577232    -0.884411  

[5 rows x 100 columns]

### Keep only certain Columns

Also while this data set is interesting for a number of different reasons, let's reduce the data that we're going to look at to just the total **population estimates** and the total **number of births**.



We can do this by creating a list of column names that we want to keep then project those and assign the resulting `DataFrame` to our `df` variable.

In [46]:
columns_to_keep = [
    'STNAME',
    'CTYNAME',
    'BIRTHS2010',
    'BIRTHS2011',
    'BIRTHS2012',
    'BIRTHS2013',
    'BIRTHS2014',
    'BIRTHS2015',
    'POPESTIMATE2010',
    'POPESTIMATE2011',
    'POPESTIMATE2012',
    'POPESTIMATE2013',
    'POPESTIMATE2014',
    'POPESTIMATE2015'
]
df = df[columns_to_keep]
df.head()

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   
4  Alabama     Bibb County          44         266         245         259   
5  Alabama   Blount County         183         744         710         646   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   
4         247         253            22861            22733            22642   
5         618         603            57373            57711            57776   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489  
4            22512            22549            22583  
5            57734            57658            57673

### MultiIndex

The US Census data breaks down estimates of population data by **state** and **county**. We can load the data and set the index to be a combination of the state and county values and see how `pandas` handles it in a `DataFrame`. We do this by creating a list of the column identifiers we want to have indexed. And then calling `set_index` with this list and assigning the output as appropriate. We see here that we have a dual index, first the state name and then the county name.

In [47]:
df = df.set_index(['STNAME', 'CTYNAME'])
df.head()

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   
        Barbour County          70         335         300         283   
        Bibb County             44         266         245         259   
        Blount County          183         744         710         646   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   
        Barbour County         260         269            27341   
        Bibb County            247         253            22861   
        Blount County          618         603            57373   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   
        Barbour County            27226            27159            26973   
        Bibb County               22733            22642            22512   
        Blount County             57711            57776            57734   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709  
        Barbour County            26815            26489  
        Bibb County               22549            22583  
        Blount County             57658            57673

An immediate question which comes up is how we can query this `DataFrame`. For instance, we saw previously that the `loc` attribute of the `DataFrame` can take multiple arguments. And it could query both the row and the columns. When you use a `MultiIndex`, you must provide the arguments in order by the level you wish to query. Inside the index, each column is called a level and the outermost column is level zero.

For example, if we want to see the population results for `Orange County`, the first argument you want to give is the state of `California`.

In [48]:
df.loc['California', 'Orange County']

BIRTHS2010            9304
BIRTHS2011           38238
BIRTHS2012           37842
BIRTHS2013           37621
BIRTHS2014           37706
BIRTHS2015           37776
POPESTIMATE2010    3017866
POPESTIMATE2011    3056084
POPESTIMATE2012    3089343
POPESTIMATE2013    3120180
POPESTIMATE2014    3144961
POPESTIMATE2015    3169776
Name: (California, Orange County), dtype: int64

You might be interested in just comparing two counties. For instance, `Orange County` and `San Francisco County`. To do this, we can pass the `loc` method, a list of tuples which describe the indices we wish to query. Since we have a `MultiIndex` of two values, the state and the county, we need to provide two values as each element of our filtering list.

In [49]:
df.loc[[('California', 'Orange County'),
        ('California', 'San Francisco County')]]

BIRTHS2010  BIRTHS2011  BIRTHS2012  \
STNAME     CTYNAME                                                    
California Orange County               9304       38238       37842   
           San Francisco County        2121        8899        8782   

                                 BIRTHS2013  BIRTHS2014  BIRTHS2015  \
STNAME     CTYNAME                                                    
California Orange County              37621       37706       37776   
           San Francisco County        9100        8844        9059   

                                 POPESTIMATE2010  POPESTIMATE2011  \
STNAME     CTYNAME                                                  
California Orange County                 3017866          3056084   
           San Francisco County           805813           816231   

                                 POPESTIMATE2012  POPESTIMATE2013  \
STNAME     CTYNAME                                                  
California Orange County                 3089343          3120180   
           San Francisco County           829517           840715   

                                 POPESTIMATE2014  POPESTIMATE2015  
STNAME     CTYNAME                                                 
California Orange County                 3144961          3169776  
           San Francisco County           852537           864816

This is how hierarchical indices work in a nutshell. They are a special part of the `pandas` library that can make management and reasoning about data easier. Of course hierarchical labeling is not just for rows. For example, you can transpose this matrix and now have hierarchical column labels. And projecting a single column which has these labels works exactly the way you would expect it to.

# Sorting and Min Max

Data exploration is an important aspect of Data Science. Before we decide do Data Analysis or Machine Learning, we must have an idea of what our data contains. The `pandas` library is equipped with a number of useful functions for this very purpose.

## Counting Values

The `value_counts()` method returns a `Series` containing the counts of unique values. This means, for any column in a `DataFrame`, this method returns the count of unique entries in that column.

The following example uses `value_counts()` to determine which country appears most often.

In [50]:
import pandas as pd

# Make example dataframe
df = pd.DataFrame(
    [
        (1, 'Germany'),
        (2, 'France'),
        (3, 'Indonesia'),
        (4, 'France'),
        (5, 'France'),
        (6, 'Germany'),
        (7, 'UK')
    ],
    columns=['groupid', 'country'],
    index=['a', 'b', 'c', 'd', 'e', 'f', 'g']
)

df

groupid    country
a        1    Germany
b        2     France
c        3  Indonesia
d        4     France
e        5     France
f        6    Germany
g        7         UK

In [51]:
# What you're looking for
values = df['country'].value_counts().keys().tolist()
counts = df['country'].value_counts().tolist()
print(df['country'].value_counts())
print("\n")
print(values)
print(counts)

country
France       3
Germany      2
Indonesia    1
UK           1
Name: count, dtype: int64


['France', 'Germany', 'Indonesia', 'UK']
[3, 2, 1, 1]


`'France'` is the country that appears most frequently with a frequency of 3 in the `DataFrame`.

`pandas`' `keys()` function returns the *info axis* for the `pandas` object. If the `pandas` object is a `Series`, it returns the index. In the given example, applied to `value_counts()`, it returns the corresponding values for the given counts.

## Sorting Index

Sorting can be done in `pandas` by using the `sort_index()` or `sort_values()` function.

`pandas`' `sort_index()` function sorts objects by labels along the given axis.
Basically the sorting algorithm is applied on the axis labels rather than the actual data in the dataframe and based on that the data is rearranged.

In [52]:
import pandas as pd

df_imdb = pd.read_csv('IMDBdata_hotlist.csv')
df_imdb.head()

Movie_Title  YR_Released  Rating  Num_Reviews   Movie_ID  \
0  The Shawshank Redemption         1994     9.2      1829037  tt0111161   
1             The Godfather         1972     9.2      1249333  tt0068646   
2    The Godfather: Part II         1974     9.0       860228  tt0071562   
3           The Dark Knight         2008     9.0      1808543  tt0468569   
4              12 Angry Men         1957     8.9       497242  tt0050083   

               Record  Runtime  
0  Top 250 Eng Movies    142.0  
1  Top 250 Eng Movies    175.0  
2  Top 250 Eng Movies    202.0  
3  Top 250 Eng Movies    152.0  
4  Top 250 Eng Movies     96.0

What are the three best reviewed movies in the database? To obtain a ranking of the best, you must be able to sort. Here we sort on the `DataFrame` directly (`inplace=True`), and we want to see the highest ranked films first, so we proceed in descending order, i.e. `ascending=False`.

In [53]:
cdf_imdb = df_imdb.copy()
cdf_imdb = cdf_imdb.set_index('Rating')
cdf_imdb.sort_index(inplace=True, ascending=False)
cdf_imdb.head()

Movie_Title  YR_Released  Num_Reviews   Movie_ID  \
Rating                                                                  
9.2     The Shawshank Redemption         1994      1829037  tt0111161   
9.2                The Godfather         1972      1249333  tt0068646   
9.0       The Godfather: Part II         1974       860228  tt0071562   
9.0              The Dark Knight         2008      1808543  tt0468569   
8.9                 12 Angry Men         1957       497242  tt0050083   

                    Record  Runtime  
Rating                               
9.2     Top 250 Eng Movies    142.0  
9.2     Top 250 Eng Movies    175.0  
9.0     Top 250 Eng Movies    202.0  
9.0     Top 250 Eng Movies    152.0  
8.9     Top 250 Eng Movies     96.0

Here we show two possibilities how the first three entries can be presented. Either with the `head()` method or with the `iloc` operator that uses slicing.

In [55]:
# Show the three best movies
cdf_imdb.head(3)

Movie_Title  YR_Released  Num_Reviews   Movie_ID  \
Rating                                                                  
9.2     The Shawshank Redemption         1994      1829037  tt0111161   
9.2                The Godfather         1972      1249333  tt0068646   
9.0       The Godfather: Part II         1974       860228  tt0071562   

                    Record  Runtime  
Rating                               
9.2     Top 250 Eng Movies    142.0  
9.2     Top 250 Eng Movies    175.0  
9.0     Top 250 Eng Movies    202.0

In [54]:
# Show the three best movies
cdf_imdb.iloc[:3]

Movie_Title  YR_Released  Num_Reviews   Movie_ID  \
Rating                                                                  
9.2     The Shawshank Redemption         1994      1829037  tt0111161   
9.2                The Godfather         1972      1249333  tt0068646   
9.0       The Godfather: Part II         1974       860228  tt0071562   

                    Record  Runtime  
Rating                               
9.2     Top 250 Eng Movies    142.0  
9.2     Top 250 Eng Movies    175.0  
9.0     Top 250 Eng Movies    202.0

The best three movies in the database are `The Shawshank Redemption` and `The Godfather` I \& II

## Sorting

In the previous section we indexed the film list according to the ranking and sorted the ranking list correspondingly.

Here we use a very similar function, `sort_values()`, that lets you sort a table by any column. The `by` parameter is the name or list of names to sort by.


In [56]:
cdf_imdb = df_imdb.copy()
cdf_imdb = cdf_imdb.sort_values(by='Rating', ascending=False)
cdf_imdb

Movie_Title  YR_Released  Rating  Num_Reviews  \
0        The Shawshank Redemption         1994     9.2      1829037   
1                   The Godfather         1972     9.2      1249333   
2          The Godfather: Part II         1974     9.0       860228   
3                 The Dark Knight         2008     9.0      1808543   
4                    12 Angry Men         1957     8.9       497242   
..                            ...          ...     ...          ...   
832                  Pledge This!         2006     1.7        15836   
833  Superbabies: Baby Geniuses 2         2004     1.6        26505   
834             Daniel the Wizard         2004     1.6        13295   
835              Saving Christmas         2014     1.5        12765   
836             Code Name: K.O.Z.         2015     1.4        24623   

      Movie_ID              Record  Runtime  
0    tt0111161  Top 250 Eng Movies    142.0  
1    tt0068646  Top 250 Eng Movies    175.0  
2    tt0071562  Top 250 Eng Movies    202.0  
3    tt0468569  Top 250 Eng Movies    152.0  
4    tt0050083  Top 250 Eng Movies     96.0  
..         ...                 ...      ...  
832  tt0417056   Bottom 250 Movies     91.0  
833  tt0270846   Bottom 250 Movies     88.0  
834  tt0421051   Bottom 250 Movies     81.0  
835  tt4009460   Bottom 250 Movies     80.0  
836  tt4458206   Bottom 250 Movies    114.0  

[837 rows x 7 columns]

## Min Max

`pandas` has  `min` and `max` functions, used to find the minimum value and maximum value of any given `DataFrame`.

In [57]:
df = pd.read_csv("subject_grades.csv")

In [58]:
df.head(4)

Physics  Biology  Maths
0       80       78     70
1       60       98     45
2       34       56     65
3       56       65     32

In [59]:
df.describe()

Physics    Biology      Maths
count  20.000000  20.000000  20.000000
mean   66.450000  71.200000  69.500000
std    22.705147  19.813871  17.276391
min    12.000000  23.000000  32.000000
25%    53.250000  56.000000  60.000000
50%    72.000000  71.500000  68.000000
75%    80.000000  82.000000  81.000000
max    98.000000  98.000000  99.000000

Show descriptive statistics only for the column of `"Maths"`:

In [60]:
df['Maths'].describe()

count    20.000000
mean     69.500000
std      17.276391
min      32.000000
25%      60.000000
50%      68.000000
75%      81.000000
max      99.000000
Name: Maths, dtype: float64

In [61]:
maths_mean = df['Maths'].mean()
physics_max = df['Physics'].max()
biology_min = df['Biology'].min()
print(maths_mean)
print(physics_max)
print(biology_min)

69.5
98
23


Calculate Min und Max per row:

In [62]:
df['Min'] = df.min(axis=1)
df['Max'] = df.max(axis=1)

In [63]:
df.head(5)

Physics  Biology  Maths  Min  Max
0       80       78     70   70   80
1       60       98     45   45   98
2       34       56     65   34   65
3       56       65     32   32   65
4       98       23     55   23   98

Calculate Min und Max only for `"Physics"` and `"Maths"` rows:

In [64]:
df['PMMin'] = df.loc[:, ['Physics', 'Maths']].min(axis=1)
df['PMMax'] = df.loc[:, ['Physics', 'Maths']].max(axis=1)

In [65]:
df.head(5)

Physics  Biology  Maths  Min  Max  PMMin  PMMax
0       80       78     70   70   80     70     80
1       60       98     45   45   98     45     60
2       34       56     65   34   65     34     65
3       56       65     32   32   65     32     56
4       98       23     55   23   98     55     98

## `idxmin()` and `idxmax()`

In a nutshell, those functions will return the ID (the index position) of the smallest `idxmin()` and largest `idxmax()` item.

For example, to return to the already mentioned reviewed movie database. Let's assume that we only want to find one of the best and one of the worst rated movies, then we can do the following:

In [66]:
import pandas as pd

cdf_imdb = pd.read_csv('IMDBdata_hotlist.csv')

In [67]:
bestindex = cdf_imdb['Rating'].idxmax()

# One of the best movies, i.e. highest rating
bestmovie = cdf_imdb.iloc[bestindex]['Movie_Title']

print("One of the best movies is «{}»".format(bestmovie))

One of the best movies is «The Shawshank Redemption»


Here we apply the function `idxmax()` to the `"Rating"` column and get the corresponding index of one of the movies with the maximum rating. This index can be used with the `iloc` operator to get the data frame row of the movie. Using operation chaining, we select the movie title column to get the movie name as a string.

In [68]:
print("One of the worst movies is «{}»".format(cdf_imdb.iloc[cdf_imdb['Rating'].idxmin()]['Movie_Title']))

One of the worst movies is «Code Name: K.O.Z.»


Here we use a comprehensive code, essentially using `idxmin` in the style of the previous `idxmax` example. This means that we use `idxmin()` to get the index of one of the movies with the lowest rating.

You can also find the function `argmax()` in source code on the Web or elsewhere. But there is not much difference between `idxmax()` and `argmax()`. `idxmax` returns the index name of the maximum value. `argmax` does the same. However, `idxmax` is a `pandas` function, so it can take `Series` and `DataFrame`. `argmax` is a `NumPy` function.

